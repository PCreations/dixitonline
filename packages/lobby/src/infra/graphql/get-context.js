export const makeGetContext = ({ dispatchDomainEvents, authorizationService } = {}) => async ({ req }) => ({
  dispatchDomainEvents,
  currentUser: await authorizationService.getCurrentUser(req),
});
