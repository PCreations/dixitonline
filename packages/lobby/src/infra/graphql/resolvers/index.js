import { makeCreateNewGame } from '../../../useCases/create-new-game';
import { makeGetGames } from '../../../useCases/get-games';
import { makeJoinGame } from '../../../useCases/join-game';
import { makeStartGame } from '../../../useCases/start-game';
import { makePlayer } from '../../../domain/player';
import {
  GameAlreadyJoinedByPlayerError,
  MaximumNumberOfPlayerReachedError,
  OnlyHostCanStartGameError,
} from '../../../domain/game';

export const resolvers = {
  Query: {
    lobbyGames(_, __, { dataSources }) {
      const getGames = makeGetGames({ lobbyRepository: dataSources.lobbyRepository });
      return getGames();
    },
  },
  Mutation: {
    async lobbyCreateGame(_, __, context) {
      const { dataSources, dispatchDomainEvents, currentUser } = context;
      const createNewGame = makeCreateNewGame({ lobbyRepository: dataSources.lobbyRepository });
      const [game, domainEvents] = await createNewGame(makePlayer({ id: currentUser.id, name: currentUser.username }));
      dispatchDomainEvents(domainEvents);
      return {
        game,
      };
    },
    async lobbyJoinGame(_, { lobbyJoinGameInput }, context) {
      const { dataSources, dispatchDomainEvents, currentUser } = context;
      const { gameId } = lobbyJoinGameInput;
      const joinGame = makeJoinGame({ lobbyRepository: dataSources.lobbyRepository });
      try {
        const [game, domainEvents] = await joinGame({ gameId, currentUser });
        dispatchDomainEvents(domainEvents);
        return {
          __typename: 'LobbyJoinGameResultSuccess',
          game,
        };
      } catch (error) {
        const knownErrors = [GameAlreadyJoinedByPlayerError, MaximumNumberOfPlayerReachedError];
        if (knownErrors.some(KnownError => error instanceof KnownError)) {
          const type =
            error instanceof GameAlreadyJoinedByPlayerError
              ? 'GAME_ALREADY_JOINED'
              : 'MAXIMUM_NUMBER_OF_PLAYERS_REACHED';
          return {
            __typename: 'LobbyJoinGameResultError',
            type,
          };
        }
        throw error;
      }
    },
    async lobbyStartGame(_, { lobbyStartGameInput }, { dataSources, dispatchDomainEvents, currentUser }) {
      const { lobbyRepository } = dataSources;
      const { gameId } = lobbyStartGameInput;
      const startGame = makeStartGame({ lobbyRepository, currentUser });
      const result = await startGame({ gameId });
      if (result.error) {
        if (result.error instanceof OnlyHostCanStartGameError) {
          return {
            __typename: 'LobbyStartGameResultError',
            type: 'ONLY_HOST_CAN_START_GAME',
          };
        }
      }
      dispatchDomainEvents(result.events);
      return {
        __typename: 'LobbyStartGameResultSuccess',
        gameId: result.value,
      };
    },
  },
  LobbyGame: {
    players: game => [game.host, ...game.players],
  },
};
