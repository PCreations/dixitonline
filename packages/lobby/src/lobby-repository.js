import { v1 as uuidv1 } from 'uuid';

export const makeLobbyRepository = ({ uuid = uuidv1, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    createGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      return doc.data();
    },
  };
};

export const makeNullLobbyRepository = ({ nextGameId, gamesData }) =>
  makeLobbyRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
              };
            },
          };
        },
      };
    },
  };
};
