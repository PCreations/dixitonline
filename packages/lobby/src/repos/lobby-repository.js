import { v1 as uuidv1 } from 'uuid';

export const makeLobbyRepository = ({ uuid = uuidv1, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    createGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      return doc.data();
    },
    getAllGames() {
      return lobbyGames.get().then(snapshot => {
        const games = [];
        snapshot.forEach(doc => games.push(doc.data()));
        return games;
      });
    },
  };
};

export const makeNullLobbyRepository = ({ nextGameId, gamesData }) =>
  makeLobbyRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
              };
            },
            set(game) {
              gamesData[gameId] = game;
            },
          };
        },
        async get() {
          const docs = Object.values(gamesData).map(game => ({
            data() {
              return game;
            },
          }));
          return {
            forEach: docs.forEach.bind(docs),
          };
        },
      };
    },
  };
};
