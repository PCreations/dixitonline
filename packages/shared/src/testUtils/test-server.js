import { ApolloServer } from 'apollo-server';
import { makeNullGraphqlExpressAuthorizationService } from '@dixit/users';

export const createMakeTestServer = ({ schema, getContext }) => ({
  getDataSources,
  dispatchDomainEvents = () => {},
  currentUserId = '',
  currentUserUsername = '',
} = {}) =>
  new ApolloServer({
    schema,
    dataSources: getDataSources,
    context: getContext({
      dispatchDomainEvents,
      authorizationService: makeNullGraphqlExpressAuthorizationService({
        userIdInDecodedToken: currentUserId,
        currentUserUsername,
      }),
    }),
  });
