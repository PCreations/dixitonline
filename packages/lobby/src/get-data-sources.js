import { makeLobbyRepository } from './lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
