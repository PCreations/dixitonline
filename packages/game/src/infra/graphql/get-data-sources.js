import { makeGameRepository } from '../../repos/game-repository';

export const makeGetDataSources = ({ gameRepository = makeGameRepository() } = {}) => () => ({
  gameRepository,
});
