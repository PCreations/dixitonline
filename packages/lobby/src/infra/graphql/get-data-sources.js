import { makeLobbyRepository } from '../../repos/lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
