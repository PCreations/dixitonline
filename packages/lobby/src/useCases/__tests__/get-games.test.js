import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import { makeNullLobbyRepository } from '../../repos/lobby-repository';
import { makeGetDataSources } from '../../infra/graphql/get-data-sources';
import { makeTestServer } from '../../testsUtils/test-server';

describe('getGames', () => {
  test('gets all the lobby games', async () => {
    // arrange
    const server = makeTestServer({
      getDataSources: makeGetDataSources({
        lobbyRepository: makeNullLobbyRepository({
          nextGameId: 'g3',
          gamesData: { g1: { id: 'g1' }, g2: { id: 'g2' } },
        }),
      }),
    });
    const LOBBY_GAMES = gql`
      query LobbyGames {
        lobbyGames {
          id
        }
      }
    `;

    // act
    const { query } = createTestClient(server);

    // assert
    const response = await query({ query: LOBBY_GAMES });
    expect(response).toMatchSnapshot();
  });
});
