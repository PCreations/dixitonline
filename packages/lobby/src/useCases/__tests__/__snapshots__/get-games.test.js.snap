// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGames gets all the lobby games 1`] = `
Object {
  "data": Object {
    "lobbyGames": Array [
      Object {
        "id": "g1",
      },
      Object {
        "id": "g2",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
